Kaggle Link: https://www.kaggle.com/code/sathwikhegde15/neural-network-type-classification-typeface-mnist


**Abstract:**

In this notebook, We will create a model that can identify handwriting digits by using TMNIST dataset.
We are trying to solve here is to classify grayscale images of handwritten digits(28 x 28), into the categories 0-9. 

This dataset we are using is the MNIST database for handwritten digits. It consists of images representing digits from 0-9 produced using 2,990 google fonts files.

Solving MNIST is like Hello World problem of Deep learning.

The dataset consists of a single file:TMNIST_Data.csv

This file consists of 29,900 examples with labels and font names.

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from tensorflow import keras
%matplotlib inline


import keras
from sklearn.model_selection import train_test_split
import torch
import torch.nn as nn
from torch.utils.data import Dataset, DataLoader, TensorDataset
from sklearn.preprocessing import LabelBinarizer
import os
os.environ['CUDA_VISIBLE_DEVICES'] = "0"
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")

# Load Data

In [2]:
# The dataset is in the form set of 4 numpy arrays
data = pd.read_csv("TMNIST_Data.csv")
X = data.drop(columns={'names', 'labels'})
y = data[['labels']]
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2)

In [3]:
y_train = y_train.values.reshape((-1,))
y_test = y_test.values.reshape((-1,))

In [4]:
# from keras.datasets import mnist
# (train_images, train_labels), (test_images, test_labels)=mnist.load_data()

We will train  the images and the labels form the training_sets, which isthe data the model will learn from. The model will then be tested on the "test_set", test_images and test_labels

In [5]:
X_train.shape

(23920, 784)

In [6]:
X_test.shape

(5980, 784)

In [7]:
y_train.shape

(23920,)

In [8]:
y_test.shape

(5980,)

We have a 28x28 pixel image sample dataset.

## Building Model

In [9]:
from keras import models
from keras import layers

model = models.Sequential()
model.add(layers.Dense(512,activation='relu', input_shape=(28*28,)))
model.add(layers.Dense(10,activation='softmax'))

Our network consists of 2 dense layers (fully connected neural layers). The second layer is softmax layer which will return an array of 10 scores.Sequential means we are have a stack of layers in our neural network. Since it is a stack it will accept every layer as an element. Here we have defined input as 784 neurons and input as 10 neurons. We are using activation function as softmax.

## Compile the Keras Model

In [10]:
model.compile(optimizer="rmsprop",
              loss="categorical_crossentropy",
              metrics=['accuracy'])

After running the model we have to compile the model.
We are using three parameters,
1) optimizer = "rmsprop". It will help to achieve gobal optimization in an efficient way.
2) loss = categorical_crossentropy. Which means, the output label is assigned one-hot category encoding value in form of 0s and 1. The output label, if present in integer form, is converted into categorical encoding using keras. 3) metrics = accuracy, since our goal is to make it more accurate.

# Preprocess Data:
We are going to flatten the data using reshape function and convert it into 1d array fom 2d array.
Then we will scale our data by dividing it by 255 to improve the accuracy. We will transform it to a shape float32, so the values are between 0 and 1.



In [11]:
X_train=X_train.values.reshape(23920,784)
X_train=X_train.astype('float32')/255

X_test=X_test.values.reshape(5980,784)
X_test=X_test.astype('float32')/255

Now we will catgorically encode the labels

In [12]:
from tensorflow.keras.utils import to_categorical

y_train=to_categorical(y_train)
y_test=to_categorical(y_test) 

We will now train the model by doing Fit on the keras model

In [13]:
result = model.fit(X_train,y_train, epochs = 5, batch_size=128)

Epoch 1/5
187/187 [==============================] - 4s 16ms/step - loss: 0.2587 - accuracy: 0.9333
Epoch 2/5
187/187 [==============================] - 3s 18ms/step - loss: 0.1083 - accuracy: 0.9705
Epoch 3/5
187/187 [==============================] - 3s 18ms/step - loss: 0.0805 - accuracy: 0.9787
Epoch 4/5
187/187 [==============================] - 3s 17ms/step - loss: 0.0633 - accuracy: 0.9826
Epoch 5/5
187/187 [==============================] - 3s 16ms/step - loss: 0.0504 - accuracy: 0.9859


After running it for 5 epochs we get an accuracy of 98%. Lets try by increasing the epochs if we get an increased accuracy. We will run it for 10 epochs.

In [14]:
result = model.fit(X_train,y_train, epochs = 10, batch_size=128)

Epoch 1/10
187/187 [==============================] - 2s 12ms/step - loss: 0.0409 - accuracy: 0.9883
Epoch 2/10
187/187 [==============================] - 2s 9ms/step - loss: 0.0333 - accuracy: 0.9908
Epoch 3/10
187/187 [==============================] - 2s 9ms/step - loss: 0.0276 - accuracy: 0.9921
Epoch 4/10
187/187 [==============================] - 2s 9ms/step - loss: 0.0224 - accuracy: 0.9941
Epoch 5/10
187/187 [==============================] - 2s 9ms/step - loss: 0.0195 - accuracy: 0.9950
Epoch 6/10
187/187 [==============================] - 2s 9ms/step - loss: 0.0155 - accuracy: 0.9957
Epoch 7/10
187/187 [==============================] - 2s 9ms/step - loss: 0.0142 - accuracy: 0.9959
Epoch 8/10
187/187 [==============================] - 2s 9ms/step - loss: 0.0119 - accuracy: 0.9970
Epoch 9/10
187/187 [==============================] - 2s 9ms/step - loss: 0.0124 - accuracy: 0.9972
Epoch 10/10
187/187 [==============================] - 2s 9ms/step - loss: 0.0094 - accuracy: 0.997

The accuracy increases to 99.77% after running it for 10 epochs.

Evaluate the Keras Model

In [15]:
train_loss, train_acc=model.evaluate(X_train, y_train)

748/748 [==============================] - 2s 2ms/step - loss: 0.0055 - accuracy: 0.9989


In [16]:
print('Train Accuracy:', train_acc)
print('Train loss:', train_loss)

Train Accuracy: 0.998913049697876
Train loss: 0.005451668053865433


We get an accuracy of 99% on the training data

Next we will check on the test data

In [17]:
test_loss, test_acc=model.evaluate(X_test, y_test)

187/187 [==============================] - 0s 3ms/step - loss: 0.0664 - accuracy: 0.9866


In [18]:
print('Test Accuracy:', test_acc)
print('Test loss:', test_loss)

Test Accuracy: 0.9866220951080322
Test loss: 0.0664043202996254


We get an accuracy of 98%

**Conclusion:**

We have a train accuracy of 98.5% when we run on 5 epochs but we get an accuracy of 99% when we increase the epoch to 10. The test accuracy of 98.66% which is just slightly less than the train accuracy, which could be a clear case of overfitting.

References:
https://www.youtube.com/watch?v=bee0GrKBCrE
https://www.youtube.com/watch?v=iqQgED9vV7k

Licensing
Copyright 2022 Sathwik Hegde

Redistribution and use in source and binary forms, with or without modification, are permitted provided that the following conditions are met:

Redistributions of source code must retain the above copyright notice, this list of conditions and the following disclaimer.

Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the following disclaimer in the documentation and/or other materials provided with the distribution.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.